File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/metadata/EntityMetadata.java
package com.foundgine.core.semantic.metadata;

import com.foundgine.core.abstractions.EntityId;
import java.util.List;

/** Static description of a domain entity. Metadata describes what exists. */
public record EntityMetadata(EntityId entityId, String name, List<ColumnMetadata> columns, String storageName,
		List<FieldMetadata> fields, ColumnReference primaryKey, Class<?> clrType, boolean event,
		ColumnReference temporalColumn, List<AliasDeclaration> aliases) {
	public EntityMetadata(EntityId entityId, String name, List<ColumnMetadata> columns) {
		this(entityId, name, columns, null, null, null, null, false, null, null);
	}

	public String effectiveStorageName() {
		return storageName != null ? storageName : name;
	}

	public List<FieldMetadata> effectiveFields() {
		return fields != null ? fields : List.of();
	}

	public List<AliasDeclaration> effectiveAliases() {
		return aliases == null ? List.of() : List.copyOf(aliases);
	}
}

File: src/csharp/Foundgine.Core/Semantic/Metadata/EntityMetadata.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Metadata;

/// <summary>
/// Static description of a domain entity. Metadata describes what exists;
/// it does not describe a request and does not know about GraphQL or SQL.
/// </summary>
public sealed record EntityMetadata(
    EntityId EntityId,
    string Name,
    IReadOnlyList<ColumnMetadata> Columns,
    string? StorageName = null,
    IReadOnlyList<FieldMetadata>? Fields = null,
    ColumnReference? PrimaryKey = null,
    Type? ClrType = null,
    // True when this entity represents an occurrence at a point in time (an
    // event) rather than the current state of something. See
    // FoundgineEventAttribute. Defaults to false (a state entity).
    bool IsEvent = false,
    // The column carrying the timestamp the event occurred at, when declared
    // and the entity IsEvent. Null for state entities and for events that did
    // not declare a temporal column.
    ColumnReference? TemporalColumn = null,
    IReadOnlyList<AliasDeclaration>? Aliases = null)
{
    public string EffectiveStorageName => StorageName ?? Name;
    public IReadOnlyList<FieldMetadata> EffectiveFields => Fields ?? [];
}